pep-0458.txt

```diff
@@ -533,101 +662,103 @@ same metadata or distributions.  There are also issues with consistency on PyPI
+lists the
+version of the *snapshot* metadata, which in turn lists the versions of the
+*targets* and delegated targets metadata, all as part of a given consistent
 snapshot.
 
+Eventually, *targets* or delegated targets metadata point to the actual target
+files, including their `cryptographic hashes`__. Thus, to mark a target file as
+part of a consistent snapshot it MUST, when written to disk, include its hash
+in its filename:
+
+  HASH.FILENAME
+    where HASH is the `hex digest`__ of the `SHA-256`__ hash of the file
+    contents and FILENAME is the original filename.
+
+__ https://en.wikipedia.org/wiki/Cryptographic_hash_function
+__ https://docs.python.org/3.7/library/hashlib.html#hashlib.hash.hexdigest
+__ https://en.wikipedia.org/wiki/SHA-2
+
+
+Assuming infinite disk space, strictly incrementing version numbers, and no
+`hash collisions`__, a client may safely read from one snapshot while PyPI
+produces another snapshot.
+
 __ https://en.wikipedia.org/wiki/Collision_(computer_science)
 
 In this simple but effective manner, PyPI is able to capture a consistent
 snapshot of all projects and the associated metadata at a given time.  The next
 subsection provides implementation details of this idea.
 
 Note: This PEP does not prohibit using advanced file systems or tools to
-produce consistent snapshots. There are two important reasons for why this PEP
-proposes the simple solution.  First, the solution does not mandate that PyPI
+produce consistent snapshots. There are two important reasons for proposing a simple solution in this PEP.
+First, the solution does not mandate that PyPI
 use any particular file system or tool.  Second, the generic file-system based
-approach allows mirrors to use extant file transfer tools such as rsync to
+approach allows mirrors to use extant file transfer tools, such as rsync, to
 efficiently transfer consistent snapshots from PyPI.
 
 
 Producing Consistent Snapshots
 ------------------------------
 
-Given a project, PyPI is responsible for updating the *bins* metadata (roles
-delegated by the *bins* role and signed with an online key).  Every project
-MUST upload its release in a single transaction.  The uploaded set of files is
-called the "project transaction".  How PyPI MAY validate the files in a project
-transaction is discussed in a later section.  For now, the focus is on how PyPI
-will respond to a project transaction.
-
-Every metadata and target file MUST include in its filename the `hex digest`__
-of its `SHA-256`__ hash.  For this PEP, it is RECOMMENDED that PyPI adopt a
-simple convention of the form: digest.filename, where filename is the original
-filename without a copy of the hash, and digest is the hex digest of the hash.
-
-__ http://docs.python.org/2/library/hashlib.html#hashlib.hash.hexdigest
-__ https://en.wikipedia.org/wiki/SHA-2
+When given a project, PyPI is responsible for updating the *bin-n* metadata.  Every
```

Comment:
If I understand this correctly, this means that for a given release, PyPI must wait for all files for that release to exist before signing metadata? That may be an issue for two reasons:

1) files are uploaded one-by-one currently
2) PyPI allows for upload of additional files for a release at any point in the future

Are these cases allowable?